synapse/lib/jsonstor.py

```diff
@@ -0,0 +1,380 @@
+        '''
+        pkey = self._pathToPkey(path)
+        buid = self.slab.pop(pkey, db=self.pathdb)
+        if buid is not None:
+            self._incRefObj(buid, valu=-1)
+
+    async def setPathLink(self, srcpath, dstpath):
+        '''
+        Add a link from the given srcpath to the dstpath.
+        NOTE: This causes the item at dstpath to be incref'd
+        '''
+        srcpkey = self._pathToPkey(srcpath)
+        dstpkey = self._pathToPkey(dstpath)
+
+        if self.slab.get(srcpkey, db=self.pathdb):
+            raise s_exc.PathExists(path=srcpath)
+
+        buid = self.slab.get(dstpkey, db=self.pathdb)
+        if buid is None:
+            raise s_exc.NoSuchPath(path=dstpath)
+
+        self._incRefObj(buid, valu=1)
+        self.slab.put(srcpkey, buid, db=self.pathdb)
+
+    async def getPathObjProp(self, path, prop):
+
+        item = await self.getPathObj(path)
+        if item is None:
+            return None
+
+        for name in self._pathToTupl(prop):
+            item = item[name]
+
+        return item
+
+    def _pathToPkey(self, path):
+        path = self._pathToTupl(path)
+        return ('\x00'.join(path)).encode()
+
+    def _pathToTupl(self, path):
+
+        if isinstance(path, str):
+            path = path.split('/')
+
+        return path
+
+    def _tuplToPath(self, path):
+        return '/'.join(path)
+
+    def _pkeyToPath(self, pkey):
+        return pkey.decode().split('\x00')
+
+    async def getPathList(self, path):
+        path = self._pathToTupl(path)
+
+        plen = len(path)
+        pkey = self._pathToPkey(path)
+
+        for lkey, buid in self.slab.scanByPref(pkey, db=self.pathdb):
+            yield self._tuplToPath(self._pkeyToPath(lkey)[plen:])
+
+    async def getPathObjs(self, path):
+        path = self._pathToTupl(path)
+
+        plen = len(path)
+        pkey = self._pathToPkey(path)
+
+        for lkey, buid in self.slab.scanByPref(pkey, db=self.pathdb):
+            yield self._pkeyToPath(lkey)[plen:], self._getBuidItem(buid)
+
+    async def setPathObjProp(self, path, prop, valu):
+
+        buid = self._pathToBuid(path)
+        if buid is None:
+            return False
+
+        item = self._getBuidItem(buid)
+        if item is None:
+            return False
+
+        step = item
+        names = self._pathToTupl(prop)
+        for name in names[:-1]:
+            step = step[name]
+
+        name = names[-1]
+        if step.get(name, s_common.novalu) == valu:
+            return True
+
+        step[name] = valu
+        self.dirty[buid] = item
+        return True
+
+    async def delPathObjProp(self, path, prop):
+
+        buid = self._pathToBuid(path)
+        if buid is None:
+            return False
+
+        item = self._getBuidItem(buid)
+        if item is None:
+            return False
+
+        step = item
+        names = self._pathToTupl(prop)
+        for name in names[:-1]:
+            step = step[name]
+
+        del step[names[-1]]
```

Comment:
Probably pop and check if it had a value instead of blowing up here unless that was intended